### PR TITLE
added support for PJ3201A 24G presence sensor (Tuya)

### PR DESCRIPTION
**Pull request by odx**
added support for PJ3201A 24G presence sensor (Tuya) discussed here: https://github.com/Koenkk/zigbee2mqtt/issues/24777#issuecomment-2509590883

### CODE REVIEW

File: src/devices/tuya.ts [hunk 1]
```diff
@@ -13146,6 +13146,102 @@ const definitions: DefinitionWithExtend[] = [
...
+            tuyaDatapoints: [
+                [104, 'presence', tuya.valueConverter.trueFalse1],
+                [112, 'occupancy', tuya.valueConverter.trueFalseInvert],
+                [9, 'target_dis_closest', tuya.valueConverter.divideBy100],
+                [101, 'mov_delay', tuya.valueConverter.raw],
+                [102, 'none_delay', tuya.valueConverter.raw],
+                [103, 'illuminance', tuya.valueConverter.divideBy10],
+                [105, 'far_mov_sensitivity', tuya.valueConverter.raw],
+                [110, 'near_mov_sensitivity', tuya.valueConverter.raw],
+                [109, 'near_occ_sensitivity', tuya.valueConverter.raw],
```
Comment (Koenkk) [inline src/devices/tuya.ts:13235]:
```suggestion
                [109, 'near_occupancy_sensitivity', tuya.valueConverter.raw],
```

File: src/devices/tuya.ts [hunk 2]
```diff
@@ -13146,6 +13146,102 @@ const definitions: DefinitionWithExtend[] = [
...
+        ],
+        meta: {
+            tuyaDatapoints: [
+                [104, 'presence', tuya.valueConverter.trueFalse1],
+                [112, 'occupancy', tuya.valueConverter.trueFalseInvert],
+                [9, 'target_dis_closest', tuya.valueConverter.divideBy100],
+                [101, 'mov_delay', tuya.valueConverter.raw],
+                [102, 'none_delay', tuya.valueConverter.raw],
+                [103, 'illuminance', tuya.valueConverter.divideBy10],
+                [105, 'far_mov_sensitivity', tuya.valueConverter.raw],
```
Comment (Koenkk) [inline src/devices/tuya.ts:13233]:
```suggestion
                [105, 'far_move_sensitivity', tuya.valueConverter.raw],
```

File: src/devices/tuya.ts [hunk 3]
```diff
@@ -13146,6 +13146,102 @@ const definitions: DefinitionWithExtend[] = [
...
+                .withDescription('The farest distance that can be detected (present)')
+                .withUnit('m'),
+            e.binary('restore_factory', ea.STATE_SET, 'ON', 'OFF').withDescription('restore_factory'),
+            e.binary('led_indicator', ea.STATE_SET, 'ON', 'OFF').withDescription('turn on or off the led '),
+        ],
+        meta: {
+            tuyaDatapoints: [
+                [104, 'presence', tuya.valueConverter.trueFalse1],
+                [112, 'occupancy', tuya.valueConverter.trueFalseInvert],
+                [9, 'target_dis_closest', tuya.valueConverter.divideBy100],
```
Comment (Koenkk) [inline src/devices/tuya.ts:13229]:
```suggestion
                [9, 'target_distance_closest', tuya.valueConverter.divideBy100],
```

File: src/devices/tuya.ts [hunk 4]
```diff
@@ -13146,6 +13146,102 @@ const definitions: DefinitionWithExtend[] = [
             ],
         },
     },
+    {
+        fingerprint: tuya.fingerprint('TS0601', ['_TZE204_eaulras5']),
+        zigbeeModel: ['TS0601'],
```
Comment (Koenkk) [inline src/devices/tuya.ts:13151]:
```suggestion
```

Not needed, `fingerprint` already covers it.

File: src/devices/tuya.ts (hunk 1 above)
Comment (odx) [reply src/devices/tuya.ts:13235]: I left the keys in place as specified by the vendor (simple excel file with the datapoint specifications). Do you prefer readability or would you recommend to stick closer to the vendor documentation?

File: src/devices/tuya.ts [hunk 5]
```diff
@@ -13334,6 +13334,101 @@ const definitions: DefinitionWithExtend[] = [
...
+            e
+                .numeric('none_delay', ea.STATE_SET)
+                .withValueMin(0)
+                .withValueMax(43200)
+                .withValueStep(1)
+                .withDescription('setting the delay time to keep present status (the time from present -> None)')
+                .withUnit('s'),
+            e.illuminance(),
+            e
+                .numeric('far_mov_sensitivity', ea.STATE_SET)
```
Comment (trinode-work) [inline src/devices/tuya.ts:13365]: I've just tested this, it's mostly working / looking good but you need to update the reference to `far_mov_sensitivity` in the exposes after you renamed the datapoint.

same with target_distance_closest / target_dis_closest not matching


image

```suggestion
                .numeric('far_move_sensitivity', ea.STATE_SET)
```

File: src/devices/tuya.ts [hunk 6]
```diff
@@ -13334,6 +13334,101 @@ const definitions: DefinitionWithExtend[] = [
...
+        meta: {
+            tuyaDatapoints: [
+                [104, 'presence', tuya.valueConverter.trueFalse1],
+                [112, 'occupancy', tuya.valueConverter.trueFalseInvert],
+                [9, 'target_distance_closest', tuya.valueConverter.divideBy100],
+                [101, 'mov_delay', tuya.valueConverter.raw],
+                [102, 'none_delay', tuya.valueConverter.raw],
+                [103, 'illuminance', tuya.valueConverter.divideBy10],
+                [105, 'far_move_sensitivity', tuya.valueConverter.raw],
+                [110, 'near_mov_sensitivity', tuya.valueConverter.raw],
```
Comment (trinode-work) [inline src/devices/tuya.ts:13421]:
```suggestion
                [110, 'near_move_sensitivity', tuya.valueConverter.raw],
```

File: src/devices/tuya.ts [hunk 7]
```diff
@@ -13334,6 +13334,101 @@ const definitions: DefinitionWithExtend[] = [
...
+                .withUnit('s'),
+            e.illuminance(),
+            e
+                .numeric('far_mov_sensitivity', ea.STATE_SET)
+                .withValueMin(0)
+                .withValueMax(10)
+                .withValueStep(1)
+                .withDescription('the moving detecting sensitivity 1 meter away'),
+            e
+                .numeric('near_mov_sensitivity', ea.STATE_SET)
```
Comment (trinode-work) [inline src/devices/tuya.ts:13371]:
```suggestion
                .numeric('near_move_sensitivity', ea.STATE_SET)
```

File: src/devices/tuya.ts [hunk 8]
```diff
@@ -13334,6 +13334,101 @@ const definitions: DefinitionWithExtend[] = [
...
+        model: 'PJ3201A',
+        vendor: 'Dongguan Pinjia Technology Co.,LTD.',
+        description: 'Human Presence Sensor',
+        fromZigbee: [tuya.fz.datapoints],
+        toZigbee: [tuya.tz.datapoints],
+        configure: tuya.configureMagicPacket,
+        exposes: [
+            e.presence(),
+            e.occupancy(),
+            e.numeric('target_dis_closest', ea.STATE).withDescription('the target distance away from the sensor').withUnit('m'),
```
Comment (trinode) [inline src/devices/tuya.ts:13348]:
```suggestion
            e.numeric('target_distance_closest', ea.STATE).withDescription('the target distance away from the sensor').withUnit('m'),
```

**Comment by trinode** [comment]
Did you resolve the illuminance reporting issue (only reports once, after a factory reset, then nevermore)?

File: src/devices/tuya.ts (hunk 5 above)
Comment (odx) [reply src/devices/tuya.ts]: Renamed the keys completely.

File: src/devices/tuya.ts (hunk 7 above)
Comment (odx) [reply src/devices/tuya.ts]: Renamed the keys completely.

**Comment by Koenkk** [comment]
Thanks!

**Pull request merged**